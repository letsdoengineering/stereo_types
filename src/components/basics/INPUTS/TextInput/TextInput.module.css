.text-input {
    border: none;
    color: var(--slate-gray);
    font-size: 1rem;
    width: 100%;
}
