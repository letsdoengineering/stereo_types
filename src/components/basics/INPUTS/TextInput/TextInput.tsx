import React, { forwardRef } from 'react'
import getClassNames from '../../../../utils/get-class-names'
import * as textInputStyles from './TextInput.module.css'

export type TextInputProps = React.ComponentPropsWithRef<'input'> & {
    id?: string | undefined
}

const TextInput = forwardRef<HTMLInputElement, TextInputProps>((props, ref) => {
    const { id, className, ...rest } = props
    const textInputClassNames = getClassNames({
        defaultClasses: [`${textInputStyles.textInput}`],
        className: className,
    })

    return <input id={id} className={textInputClassNames} type='text' ref={ref} {...rest} />
})
TextInput.displayName = 'TextInput'

export default TextInput
