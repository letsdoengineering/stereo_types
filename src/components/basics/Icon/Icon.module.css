.icon-size-XS {
    font-size: 0.875rem;
}

.icon-size-S {
    font-size: 1rem;
}

.icon-size-M {
    font-size: 1.25rem;
}

.icon-size-L {
    font-size: 1.5rem;
}

.icon-size-XL {
    font-size: 2.75rem;
}

.icon-color-blue {
    color: var(--light-blue)
}

.icon-color-grey {
    color: var(--dark-gray)
}
