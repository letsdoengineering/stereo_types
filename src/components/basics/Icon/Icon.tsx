import React from 'react'
import getClassNames from '../../../utils/get-class-names'
import iconStyles from './Icon.module.css'

export type IconProps = React.AllHTMLAttributes<HTMLSpanElement> & {
  /** string that represents a specific icon */
  icon:
    | 'search'
    | 'calendar-day'
    | 'chevron-down'
    | 'chevron-up'
    | 'chevron-left'
    | 'chevron-right'
    | 'info-circle'
    | 'moon'
    | 'map-marker-alt'
    | 'ship'
    | 'times'
    | 'exclamation-triangle'
  /** string that represents the size of the icon - applied from css */
  iconSize?: ElementSize
  /** string that represents the color of the icon applied as css*/
  iconColor?: 'Blue' | 'Grey' | 'Red'
}

/** Icon: renders a span with a class-name that applies a specific icon from the fa stylesheet */
const Icon: React.FC<IconProps> = ({
  icon,
  iconSize = 'L',
  iconColor = 'Blue',
  className,
}: IconProps) => {
  const sizeClass = iconStyles[`iconSize${iconSize}`]
  const colorClass = iconStyles[`iconColor${iconColor}`]
  const classNames = getClassNames({
    defaultClasses: ['fas', `fa-${icon}`, sizeClass, colorClass],
    className,
  })

  return <span aria-hidden='true' className={classNames} />
}

export default Icon
