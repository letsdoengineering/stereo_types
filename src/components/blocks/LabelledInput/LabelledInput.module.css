.container {
    text-align: left;
    padding: 0.5rem 0 0.5rem 1rem;
    border: 1px solid var(--outline-gray);
}

.label-text {
    margin-bottom: 0.75rem;
    display: block;
    font-weight: 700;
    color: var(--dark-gray);
}

.icon-and-input {
    display: flex;
}

.icon-and-input-reverse {
    display: flex;
    flex-direction: row-reverse;
}

.icon {
    margin-right: 1rem;
}

.icon-reverse {
    margin-right: 1rem;
    margin-left: 1rem;
}
