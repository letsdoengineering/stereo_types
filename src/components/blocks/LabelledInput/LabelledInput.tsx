import React, { forwardRef } from 'react'
import getClassNames from '../../../utils/get-class-names'
import * as labeledInputStyles from './LabelledInput.module.css'

type LabelledInputProps = {
    htmlFor: string | undefined
    label: string
    /** show an icon, maybe even provide the name of icon to show? */
    icon?: React.ReactElement
    children: React.ReactElement
    iconOnRight?: boolean
}

/** LabelledInput: component that returns a label accessibly linked to the passed in Input */
const LabelledInput = forwardRef<HTMLInputElement, LabelledInputProps>((props, ref) => {
    const { htmlFor, label, icon, iconOnRight = false, children, ...rest } = props

    const childrenProps = {
        ...rest,
        id: htmlFor,
        ref,
    }

    const childrenWithProps = React.Children.map(children, (child) => {
        // Checking isValidElement is the safe way and avoids a typescript error too.
        if (React.isValidElement(child)) {
            return React.cloneElement(child, { ...childrenProps })
        }
        return child
    })
    const iconClassNames = getClassNames({
        conditionalClasses: {
            [`${labeledInputStyles.icon}`]: !!icon && !iconOnRight,
            [`${labeledInputStyles.iconReverse}`]: iconOnRight,
        },
    })

    const iconAndInputClassNames = getClassNames({
        defaultClasses: [`${labeledInputStyles.iconAndInput}`],
        conditionalClasses: {
            [`${labeledInputStyles.iconAndInputReverse}`]: iconOnRight,
        },
    })

    return (
        <div className={labeledInputStyles.container}>
            <label htmlFor={htmlFor}>
                <span className={labeledInputStyles.labelText}>{label}</span>
                <div className={iconAndInputClassNames}>
                    {icon && <span className={iconClassNames}>{icon}</span>}
                    {childrenWithProps}
                </div>
            </label>
        </div>
    )
})

LabelledInput.displayName = 'LabelledTextInput'

export default LabelledInput
