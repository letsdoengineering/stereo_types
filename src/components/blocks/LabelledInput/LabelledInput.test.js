import React from 'react'
import LabelledInput from './LabelledInput'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import { axe } from 'jest-axe'
import TextInput from '../../basics/INPUTS/TextInput/TextInput'
import Icon from '../../basics/Icon/Icon'

describe('LabelledTextInput', () => {
    it('renders labelled input when valid element passed', () => {
        render(
            <LabelledInput id='some-input' label='some label'>
                <TextInput />
            </LabelledInput>
        )

        expect(screen.getByLabelText('some label')).toBeInTheDocument()
    })

    it('passes children through without adding id for the label htmlFor when they are not valid react elements', () => {
        render(
            <LabelledInput id='some-input' label='some label'>
                something that is not a child
            </LabelledInput>
        )

        expect(screen.queryByLabelText('some label')).not.toBeInTheDocument()
    })

    it('has not changed the JSX layout of the page from previous snapshot', () => {
        const tree = renderer
            .create(
                <LabelledInput
                    id='some-input'
                    label='some label'
                    icon={<Icon icon='calendar-day' />}
                >
                    <TextInput />
                </LabelledInput>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const { container } = render(
            <LabelledInput id='some-input' label='some label'>
                <TextInput />
            </LabelledInput>
        )
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
