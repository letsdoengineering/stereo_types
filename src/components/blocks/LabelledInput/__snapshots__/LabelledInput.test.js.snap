// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelledTextInput has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  className="container"
>
  <label>
    <span
      className="labelText"
    >
      some label
    </span>
    <div
      className="iconAndInput"
    >
      <span
        className="icon"
      >
        <span
          aria-hidden="true"
          className="fas fa-calendar-day iconSizeL iconColorBlue"
        />
      </span>
      <input
        className="textInput"
        type="text"
      />
    </div>
  </label>
</div>
`;
